// @redocly/openapi-core/lib/config/load.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConfig = exports.getConfig = exports.findConfig = exports.CONFIG_FILE_NAMES = exports.loadConfig = void 0;
const fs = require("fs");
const path = require("path");
const redocly_1 = require("../redocly");
const utils_1 = require("../utils");
const js_yaml_1 = require("../js-yaml");
const utils_2 = require("./utils");
const config_resolvers_1 = require("./config-resolvers");
const bundle_1 = require("../bundle");
const resolve_1 = require("../resolve");
const env_1 = require("../env");
const domains_1 = require("../redocly/domains");
function addConfigMetadata({ rawConfig, customExtends, configPath, tokens, files, region, externalRefResolver, }) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        if (customExtends !== undefined) {
            rawConfig.styleguide = rawConfig.styleguide || {};
            rawConfig.styleguide.extends = customExtends;
        }
        else if ((0, utils_1.isEmptyObject)(rawConfig)) {
            rawConfig.styleguide = { extends: ['recommended'], recommendedFallback: true };
        }
        if (tokens === null || tokens === void 0 ? void 0 : tokens.length) {
            if (!rawConfig.resolve)
                rawConfig.resolve = {};
            if (!rawConfig.resolve.http)
                rawConfig.resolve.http = {};
            rawConfig.resolve.http.headers = [...((_a = rawConfig.resolve.http.headers) !== null && _a !== void 0 ? _a : [])];
            for (const item of tokens) {
                const domain = domains_1.DOMAINS[item.region];
                rawConfig.resolve.http.headers.push({
                    matches: `https://api.${domain}/registry/**`,
                    name: 'Authorization',
                    envVariable: undefined,
                    value: item.token,
                }, 
                //support redocly.com domain for future compatibility
                ...(item.region === 'us'
                    ? [
                        {
                            matches: `https://api.redoc.ly/registry/**`,
                            name: 'Authorization',
                            envVariable: undefined,
                            value: item.token,
                        },
                    ]
                    : []));
            }
        }
        return (0, config_resolvers_1.resolveConfig)({
            rawConfig: Object.assign(Object.assign({}, rawConfig), { files: files !== null && files !== void 0 ? files : rawConfig.files, region: region !== null && region !== void 0 ? region : rawConfig.region }),
            configPath,
            externalRefResolver,
        });
    });
}
function loadConfig(options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { configPath = findConfig(), customExtends, processRawConfig, files, region, externalRefResolver, } = options;
        const rawConfig = yield getConfig({ configPath, processRawConfig, externalRefResolver });
        const redoclyClient = env_1.isBrowser ? undefined : new redocly_1.RedoclyClient();
        const tokens = redoclyClient && redoclyClient.hasTokens() ? redoclyClient.getAllTokens() : [];
        return addConfigMetadata({
            rawConfig,
            customExtends,
            configPath,
            tokens,
            files,
            region,
            externalRefResolver,
        });
    });
}
exports.loadConfig = loadConfig;
exports.CONFIG_FILE_NAMES = ['redocly.yaml', 'redocly.yml', '.redocly.yaml', '.redocly.yml'];
function findConfig(dir) {
    var _a;
    if (!((_a = fs === null || fs === void 0 ? void 0 : fs.hasOwnProperty) === null || _a === void 0 ? void 0 : _a.call(fs, 'existsSync')))
        return;
    const existingConfigFiles = exports.CONFIG_FILE_NAMES.map((name) => dir ? path.resolve(dir, name) : name).filter(fs.existsSync);
    if (existingConfigFiles.length > 1) {
        throw new Error(`
      Multiple configuration files are not allowed.
      Found the following files: ${existingConfigFiles.join(', ')}.
      Please use 'redocly.yaml' instead.
    `);
    }
    return existingConfigFiles[0];
}
exports.findConfig = findConfig;
function getConfig(options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { configPath = findConfig(), processRawConfig, externalRefResolver = new resolve_1.BaseResolver(), } = options;
        if (!configPath)
            return {};
        try {
            const { document, resolvedRefMap } = yield (0, config_resolvers_1.resolveConfigFileAndRefs)({
                configPath,
                externalRefResolver,
            });
            if (typeof processRawConfig === 'function') {
                yield processRawConfig(document, resolvedRefMap);
            }
            const bundledConfig = yield (0, bundle_1.bundleConfig)(document, resolvedRefMap);
            return (0, utils_2.transformConfig)(bundledConfig);
        }
        catch (e) {
            if (e instanceof utils_2.ConfigValidationError) {
                throw e;
            }
            throw new Error(`Error parsing config file at '${configPath}': ${e.message}`);
        }
    });
}
exports.getConfig = getConfig;
function createConfig(config, options) {
    return __awaiter(this, void 0, void 0, function* () {
        return addConfigMetadata(Object.assign({ rawConfig: (0, utils_2.transformConfig)(typeof config === 'string' ? (0, js_yaml_1.parseYaml)(config) : config) }, options));
    });
}
exports.createConfig = createConfig;
